core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -390,7 +390,8 @@ class ParallelFor< FunctorType
   void execute() const
     {
       const int nwork = m_policy.end() - m_policy.begin();
-      const dim3 block(  1 , CudaTraits::WarpSize * cuda_internal_maximum_warp_count(), 1);
+      const int block_size = Kokkos::Impl::cuda_get_opt_block_size< ParallelFor >( m_functor , 1, 0 , 0 );
```

Comment:
Did this function already exist and was just unused?

Also, should this affect performance of applications significantly?

Reply:
This function did exist but was unused for the 1D range policy. It is used for the TeamPolicy when AUTO is provided as TeamSize. Yes this has potential impact on application performance, but even more important it has potential impact on actual make code function. The performance impact is that many kernels will now run with 128 threads instead of 256 per block, since the opt_block_size function chooses the smallest block size possible which still allows for maximum possible occupancy. The reasoning behind this is that that would give the scheduler more flexibility in issuing blocks. The potential drawback is that a bigger block tends to have more contiguous memory access from the same SM and hence make the pre fetcher work better. 

On the benchmark suite we miniFE sees a small (~1%) negative impact on a K40, bytes_and_flops runs 1-2% faster, miniMD runs 5% and 9% faster for its two tests. On average I believe this change will have a net positive effect on codes, which is a combination of the downside effects being smaller than the upside effects, even though I would expect roughly equal number of kernels being negative effected as positive. Its just that the positivly effected kernels will gain more than the negatively effected ones will loose.

Reply:
@crtrott thanks for the details, its good to have numbers on the miniapps.